k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -105,270 +103,277 @@ class SingleNsNamer(
 
-  case class Svc(endpoints: Set[Endpoint], ports: Map[String, Int])
+  protected type PortMap = Map[String, Int]
+  protected type NumberedPortMap = Map[Int, String]
 
-  private[this] def getEndpoints(subsets: Option[Seq[v1.EndpointSubset]]): Set[Endpoint] = {
-    val endpoints = mutable.Set.empty[Endpoint]
-    for {
-      subset <- subsets.getOrElse(Seq.empty)
-      addresses <- subset.addresses
-      addrs <- addresses
-    } {
-      endpoints += Endpoint(InetAddress.getByName(addrs.ip), addrs.nodeName)
-    }
-    endpoints.toSet
-  }
+  protected case class Endpoint(ip: InetAddress, nodeName: Option[String])
 
-  private[this] def getPorts(subsets: Option[Seq[v1.EndpointSubset]]): Map[String, Int] = {
-    val portSet = mutable.Map.empty[String, Int]
-    for {
-      subset <- subsets.getOrElse(Seq.empty)
-      ports <- subset.ports
-      port <- ports
-      name <- port.name
-    } {
-      val proto = port.protocol.map(_.toUpperCase).getOrElse("TCP")
-      if (proto == "TCP") {
-        portSet(name) = port.port
-      }
-    }
-    portSet.toMap
+  protected object Endpoint {
+    def apply(addr: v1.EndpointAddress): Endpoint =
+      Endpoint(InetAddress.getByName(addr.ip), addr.nodeName)
   }
 
-  case class SvcCache(name: String, init: Svc) {
-
-    private[this] val endpointsState = Var[Set[Endpoint]](init.endpoints)
-    private[this] val portsState = Var[Map[String, Int]](init.ports)
-
-    def port(portName: String): Var[Option[Var[Addr]]] = {
-      portsState.map { portMap =>
-        val portNumber = portMap.get(portName)
-        portNumber.map(port)
-      }
-    }
-
-    def port(portNumber: Int): Var[Addr] =
-      endpointsState.map { endpoints =>
-        val addrs: Set[Address] = endpoints.map { endpoint =>
-          val isa = new InetSocketAddress(endpoint.ip, portNumber)
-          Address.Inet(isa, endpoint.nodeName.map(Metadata.nodeName -> _).toMap)
+  private[EndpointsNamer] case class EndpointsCache(
+    nsName: String,
+    serviceName: String,
+    endpoints: Set[Endpoint],
+    ports: PortMap
+  ) {
+
+    def lookupNumberedPort(
+      mappings: NumberedPortMap,
+      portNumber: Int
+    ): Option[Set[Address]] =
+      mappings.get(portNumber)
+        .flatMap { targetPort =>
+          // target may be an int (port number) or string (port name)
+          Try(targetPort.toInt).toOption match {
+            case Some(targetPortNumber) =>
+              // target port is a number and therefore exists
+              Some(port(targetPortNumber))
+            case None =>
+              // target port is a name and may or may not exist
+              port(targetPort)
+          }
         }
-        Addr.Bound(addrs)
-      }
-
-    def ports: Var[Map[String, Int]] = portsState
 
-    def update(subsets: Option[Seq[v1.EndpointSubset]]): Unit = {
-      val newEndpoints = getEndpoints(subsets)
-      val newPorts = getPorts(subsets)
+    def port(portName: String): Option[Set[Address]] =
```

Comment:
I went back and forth on that for a while and ended up using the name from the previous code. Can definitely rename it.